File: benchmark/primitive_ops/canisters/azle/data_types/int/int8.ts
import { ic, int8, nat32, nat64, Update } from 'azle';

let int8_init_heap_storage: { [key: string]: int8 | undefined; } = {};

export function int8_init_stack(num_inits: nat32): Update<nat64> {
    let i = 0;

    while (i < num_inits) {
        let value: int8 = i % 2 === 0 ? 127 : 0;
        console.log(value);
        i += 1;
    }

    return ic.performance_counter(0);
}

export function int8_init_heap(num_inits: nat32): Update<nat64> {
    let i = 0;

    while (i < num_inits) {
        int8_init_heap_storage[`element${i}`] = i % 2 === 0 ? 127 : 0;
        i += 1;
    }

    return ic.performance_counter(0);
}

File: benchmark/primitive_ops/canisters/rust/src/data_types/int/int8.rs
use std::collections::HashMap;
use std::cell::RefCell;

thread_local! {
    static INT8_INIT_HEAP_STORAGE_REF_CELL: RefCell<Int8InitHeapStorage> = RefCell::default();
}

type Int8InitHeapStorage = HashMap<String, i8>;

#[ic_cdk_macros::update]
pub fn int8_init_stack(num_inits: u32) -> u64 {
    let mut i = 0;

    while i < num_inits {
        let value: i8 = if i % 2 == 0 { 127 } else { 0 };
        bencher::black_box(&value); // Trying to ensure that the value assignment above is not optimized away
        i += 1;
    }

    ic_cdk::api::call::performance_counter(0)
}

#[ic_cdk_macros::update]
pub fn int8_init_heap(num_inits: u32) -> u64 {
    INT8_INIT_HEAP_STORAGE_REF_CELL.with(|int8_init_heap_storage_ref_cell| {
        let mut i = 0;
        let mut int8_init_heap_storage = int8_init_heap_storage_ref_cell.borrow_mut();

        while i < num_inits {
            int8_init_heap_storage.insert(
                format!("element{}", i),
                if i % 2 == 0 { 127 } else { 0 }
            );
            
            i += 1;
        }
    });

    ic_cdk::api::call::performance_counter(0)
}